### DISCUSSION TITLE
How can I query multiple Postgres+pgvector vector stores using SQLAlchemy?

### DISCUSSION THREAD

**Discussion by chrispy-snps** [sqlalchemy/sqlalchemy#11916] [Usage Questions] [answered]
I have multiple Postgres+pgvector vector store tables. For example, let's say I have `table1` and `table2` with `id`, `txt`, and `v` (vector) columns:

```python
from sqlalchemy import create_engine, MetaData, Table, Column, Integer, String, text, select, union_all
from pgvector.sqlalchemy import Vector

# create our connection
engine = create_engine(...)
metadata = MetaData()

# create two pgvector tables
with engine.begin() as conn:
    conn.execute(text("CREATE EXTENSION IF NOT EXISTS vector"))
with engine.begin() as conn:
    conn.execute(text("DROP TABLE IF EXISTS table1"))
with engine.begin() as conn:
    conn.execute(text("DROP TABLE IF EXISTS table2"))

with engine.begin() as conn:
    table1 = Table("table1", metadata, Column("id", Integer), Column("txt", String), Column("v", Vector))
    table2 = Table("table2", metadata, Column("id", Integer), Column("txt", String), Column("v", Vector))
    metadata.create_all(engine)

# push some "documents" into both tables
with engine.begin() as conn:
    conn.execute(
        table1.insert(),
        [
            {"id": 1, "txt": "aaa", "v": [0.1, 0.1]},
            {"id": 2, "txt": "bbb", "v": [0.2, 0.2]},
        ],
    )
    conn.execute(
        table2.insert(),
        [
            {"id": 3, "txt": "ccc", "v": [0.3, 0.3]},
            {"id": 4, "txt": "ddd", "v": [0.4, 0.4]},
        ],
    )
```

I want to use SQLAlchemy to query them, to make it easier to dynamically choose which vector stores to query and what columns to return at runtime. How can I do this?

**Comment by chrispy-snps** [comment, marked as answer]
In SQLAlchemy, a union of vector stores can be queried using raw SQL as follows:

```python
# query the union of tables by vector similarity in raw SQL
query_vector = [0.16, 0.16]
with engine.connect() as conn:
    result = conn.execute(
        text("""
SELECT id, txt FROM (
 SELECT id, txt, v FROM public.table1 UNION
 SELECT id, txt, v FROM public.table2
) AS U
 ORDER BY v <-> cast(:query_vector AS vector);
"""),
        {"query_vector": query_vector},
    )

# show results
for row in result:
    print(row)
```

Similarly, a union of vector stores can be queried using SQLAlchemy Core as follows:

```python
# query the union of tables by vector similarity in SQLAlchemy Core
query_vector = [0.16, 0.16]
with engine.connect() as conn:
    # get union of table results
    table_select_statements = [
        select(table1.c["id", "txt", "v"]),
        select(table2.c["id", "txt", "v"]),
    ]
    subquery = union_all(*table_select_statements).subquery()
    stmt = select(subquery.c["id", "txt"]).order_by(subquery.c["v"].l2_distance(query_vector).asc())
    result = conn.execute(stmt)

# show results
for row in result:
    print(row)
```

In the preceding example,

* The tables to query can be dynamically added to the `table_select_statements` list.
* The vector field to use for similarity search (e.g. to support multi-model vector stores) can be dynamically specified in the inner `subquery.c[]` reference within the `order_by()` statement.
* The final set of columns to return can be dynamically specified in the outer `subquery.c[]` reference assigned to `stmt`.

**Reply by CaselIT** [reply]
is there a question there of you found a solution?

**Reply by chrispy-snps** [reply]
This is something that took me some time to figure out, so I wanted to contribute my findings to the community to help others.

Feedback is welcomed on the following:

* Code simplifications or improvements
  * Can the subquery sequence be simplified?
  * Did I use the right subquery/statement terminology?
* A minimum required SQLAlchemy version that should be mentioned
  * I ran into deprecation messages during my investigation, which led me to the SQLAlchemy 1.4 What's New topic on subqueries.
* A SQLAlchemy ORM equivalent of the example
  * I don't know/use ORM, but I would be happy to add an example if provided.

**Reply by CaselIT** [reply]
I think you don't need the second subquery, or at least it's not in your raw sql.
This would also work:
```py
query_vector = [0.16, 0.16]
with engine.connect() as conn:
    # get union of table results
    table_select_statements = [
        select(table1.c["id", "txt", "v"]),
        select(table2.c["id", "txt", "v"]),
    ]
    subquery1 = union_all(*table_select_statements).subquery()
    stmt = select(subquery1.c["id", "txt"]).order_by(subquery1.c["v"].l2_distance(query_vector).asc())
    result = conn.execute(stmt)
```
